File: crates/srn-node/src/scheduler.rs
//! Residue-driven scheduler — §6 of the scheduling-mechanism paper.
//!
//! Priority rule (Definition: Residue priority):
//!   P(τ) = Δ_τ / max(ρ_τ − θ_τ, β)
//!
//! where β > 0 is the entropic floor, Δ_τ is the descent rate over a window,
//! ρ_τ is the current residue, and θ_τ is the sufficiency threshold.
//!
//! Invariants proved in the paper:
//!   • A stalled task (Δ = 0) gets P = 0 → never dispatched while others run.
//!   • A task at θ gets P = +∞ → always dispatched first (finished immediately).
//!   • Trajectory count M is strictly monotone across non-declining ticks.

use std::collections::BinaryHeap;
use std::cmp::Ordering;
use serde::{Deserialize, Serialize};

pub const FLOOR: f64 = 0.01;   // β: entropic floor (strictly positive)
pub const STALL_WINDOW: usize = 5; // w: stall declared after this many flat units

/// State of a live task in the scheduler.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum TaskState {
    Running,
    Stalled,
    Sufficient,
    Done,
    Declined,
}

/// A live task record.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Task {
    pub id: String,
    /// Categorical complexity K — number of distinguishable work-trajectories needed.
    pub complexity: u64,
    /// Operator type count d (for T(n,d) = d(d+1)^{n-1}).
    pub op_types: u64,
    /// Committed unit count M — strictly monotone.
    pub trajectory_count: u64,
    /// Current residue ρ ∈ [β, 100].
    pub residue: f64,
    /// Sufficiency threshold θ ∈ [β, 100].
    pub threshold: f64,
    /// Recent residue readings (for descent-rate window).
    residue_history: Vec<f64>,
    /// Current task state.
    pub state: TaskState,
}

impl Task {
    pub fn new(id: impl Into<String>, complexity: u64, op_types: u64, initial_residue: f64) -> Self {
        let id = id.into();
        let threshold = FLOOR; // default: run to floor; caller may override
        Self {
            id,
            complexity,
            op_types,
            trajectory_count: 0,
            residue: initial_residue,
            threshold,
            residue_history: vec![initial_residue],
            state: TaskState::Running,
        }
    }

    pub fn with_threshold(mut self, theta: f64) -> Self {
        self.threshold = theta.max(FLOOR);
        self
    }

    /// Descent rate Δ over the stall window — non-negative by definition.
    pub fn descent_rate(&self) -> f64 {
        let h = &self.residue_history;
        if h.len() < 2 {
            return 0.0;
        }
        let window = STALL_WINDOW.min(h.len() - 1);
        let oldest = h[h.len() - 1 - window];
        let newest = *h.last().unwrap();
        (oldest - newest).max(0.0) / window as f64
    }

    /// Residue priority P(τ) — the scheduling signal.
    pub fn priority(&self) -> f64 {
        if self.state != TaskState::Running {
            return 0.0;
        }
        if self.residue <= self.threshold {
            return f64::INFINITY; // finish immediately
        }
        let delta = self.descent_rate();
        if delta == 0.0 {
            return 0.0; // stalled — do not feed
        }
        let denom = (self.residue - self.threshold).max(FLOOR);
        delta / denom
    }

    /// Record a new residue reading after one committed unit.
    /// Returns the updated priority.
    pub fn commit_unit(&mut self, new_residue: f64) -> f64 {
        debug_assert!(
            new_residue >= FLOOR || new_residue <= self.residue,
            "residue may not increase"
        );
        self.trajectory_count += 1; // strictly monotone — never decremented
        self.residue = new_residue.max(FLOOR);
        self.residue_history.push(self.residue);

        // State transitions
        if self.residue <= self.threshold {
            self.state = TaskState::Sufficient;
        } else if self.descent_rate() == 0.0
            && self.residue_history.len() > STALL_WINDOW
        {
            let flat = self.residue_history
                .iter()
                .rev()
                .take(STALL_WINDOW)
                .all(|&r| (r - self.residue).abs() < 1e-9);
            if flat {
                self.state = TaskState::Stalled;
            }
        }
        self.priority()
    }

    /// Expected termination unit count: n_term(K, d) = 1 + ⌈log_{d+1}(K/d)⌉
    pub fn termination_bound(&self) -> u64 {
        let k = self.complexity as f64;
        let d = self.op_types as f64;
        if d <= 0.0 || k <= 0.0 { return 1; }
        let inner = (k / d).log(d + 1.0);
        1 + inner.ceil() as u64
    }

    /// True if task is behind its static termination prediction.
    pub fn is_behind_curve(&self) -> bool {
        self.trajectory_count > self.termination_bound()
            && self.state == TaskState::Running
    }
}

/// A task entry in the priority queue — ordered by P(τ).
#[derive(Debug)]
struct PQEntry {
    priority: f64,   // f64 doesn't impl Ord; we use a wrapper
    task_id: String,
}

impl PartialEq for PQEntry {
    fn eq(&self, other: &Self) -> bool { self.task_id == other.task_id }
}
impl Eq for PQEntry {}

impl Ord for PQEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        self.priority.partial_cmp(&other.priority)
            .unwrap_or(Ordering::Equal)
    }
}
impl PartialOrd for PQEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

/// Tick result for one dispatched unit.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TickResult {
    pub task_id: String,
    pub trajectory_count: u64,
    pub residue_before: f64,
    pub residue_after: f64,
    pub priority_after: f64,
    pub state: TaskState,
}

/// The residue-driven scheduler.
pub struct Scheduler {
    pub tasks: std::collections::HashMap<String, Task>,
}

impl Scheduler {
    pub fn new() -> Self {
        Self { tasks: std::collections::HashMap::new() }
    }

    pub fn add_task(&mut self, task: Task) {
        self.tasks.insert(task.id.clone(), task);
    }

    /// One scheduler tick over budget B (max dispatches this tick).
    ///
    /// Implements Algorithm 1 (Tick) from the scheduling-mechanism paper.
    /// Returns the list of committed results, in dispatch order.
    pub fn tick(&mut self, budget: usize, dispatch: &mut dyn FnMut(&Task) -> f64) -> Vec<TickResult> {
        let mut results = Vec::new();
        let mut remaining = budget;

        while remaining > 0 {
            // Build priority queue from running tasks
            let mut pq: BinaryHeap<PQEntry> = BinaryHeap::new();
            for task in self.tasks.values() {
                if task.state == TaskState::Running {
                    pq.push(PQEntry {
                        priority: task.priority(),
                        task_id: task.id.clone(),
                    });
                }
            }

            let top = match pq.pop() {
                Some(e) => e,
                None => break, // no runnable tasks
            };

            if top.priority == 0.0 {
                // All runnable tasks are stalled — decline
                break;
            }

            // Dispatch the highest-priority task
            let task = self.tasks.get(&top.task_id).unwrap();
            let residue_before = task.residue;

            let new_residue = dispatch(task); // caller-supplied work unit

            let task = self.tasks.get_mut(&top.task_id).unwrap();
            let priority_after = task.commit_unit(new_residue);

            results.push(TickResult {
                task_id: top.task_id.clone(),
                trajectory_count: task.trajectory_count,
                residue_before,
                residue_after: task.residue,
                priority_after,
                state: task.state.clone(),
            });

            remaining -= 1;
        }

        results
    }

    pub fn running_count(&self) -> usize {
        self.tasks.values().filter(|t| t.state == TaskState::Running).count()
    }

    pub fn stalled_tasks(&self) -> Vec<&Task> {
        self.tasks.values().filter(|t| t.state == TaskState::Stalled).collect()
    }
}

impl Default for Scheduler {
    fn default() -> Self { Self::new() }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn priority_zero_for_stalled() {
        let mut t = Task::new("t", 100, 4, 50.0);
        // force a flat residue window
        for _ in 0..=STALL_WINDOW {
            t.commit_unit(50.0);
        }
        // After enough flat readings, priority should be 0 or state stalled
        assert!(t.priority() == 0.0 || t.state == TaskState::Stalled);
    }

    #[test]
    fn trajectory_count_strictly_monotone() {
        let mut t = Task::new("t", 100, 4, 80.0);
        for i in 1..=20u64 {
            t.commit_unit((80.0 - i as f64).max(FLOOR));
            assert_eq!(t.trajectory_count, i);
        }
    }

    #[test]
    fn sufficient_at_threshold() {
        let mut t = Task::new("t", 100, 4, 80.0).with_threshold(5.0);
        t.commit_unit(3.0); // below threshold
        assert_eq!(t.state, TaskState::Sufficient);
    }

    #[test]
    fn termination_bound_logarithmic() {
        // n_term(K,d) = 1 + ceil(log_{d+1}(K/d))
        // For K=10^12, d=3: log_4(10^12/3) ≈ 19.14 → ceil = 20 → n_term = 21
        let t = Task::new("t", 1_000_000_000_000, 3, 50.0);
        let bound = t.termination_bound();
        assert_eq!(bound, 21, "expected n_term(10^12, 3) = 21, got {bound}");

        // Verify logarithmic scaling: n_term grows as O(log K)
        let t2 = Task::new("t2", 1_000_000, 3, 50.0); // K = 10^6
        let b2 = t2.termination_bound();
        assert!(b2 < bound, "n_term should be smaller for smaller K");
        assert!(b2 > 1, "n_term > 1 for non-trivial K");
    }
}

File: ts/src/scheduler.ts
/**
 * Residue-driven scheduler (network-yield §liveness; scheduling-mechanism paper §6).
 *
 * Priority rule (residue priority):
 *   P(t) = Delta_t / max(rho_t - theta_t, beta)
 * where beta > 0 is the entropic floor, Delta_t is the descent rate over a
 * window, rho_t is the current residue, and theta_t is the sufficiency threshold.
 *
 * Invariants (proved in the paper, mirrored in tests):
 *   - a stalled task (Delta = 0) gets P = 0 -> never dispatched while others run;
 *   - a task at theta gets P = +inf -> dispatched first (finishes immediately);
 *   - trajectory count M is strictly monotone across committed units;
 *   - liveness: every live task's residue reaches theta in finite time
 *     (settling bound n_term = 1 + ceil(log_{d+1}(K/d))).
 *
 * Ported from `crates/srn-node/src/scheduler.rs`.
 */

/** Entropic floor beta (strictly positive). */
export const FLOOR = 0.01;
/** Stall declared after this many flat committed units. */
export const STALL_WINDOW = 5;

export type TaskState = "running" | "stalled" | "sufficient" | "done" | "declined";

/** A live scheduler task. */
export class Task {
  readonly id: string;
  /** Categorical complexity K — distinguishable work-trajectories needed. */
  readonly complexity: number;
  /** Operator type count d (for the T(n,d) inflation / termination bound). */
  readonly opTypes: number;
  /** Committed unit count M — strictly monotone. */
  trajectoryCount = 0;
  /** Current residue rho in [beta, 100]. */
  residue: number;
  /** Sufficiency threshold theta in [beta, 100]. */
  threshold: number;
  state: TaskState = "running";
  private readonly residueHistory: number[];

  constructor(id: string, complexity: number, opTypes: number, initialResidue: number) {
    this.id = id;
    this.complexity = complexity;
    this.opTypes = opTypes;
    this.residue = initialResidue;
    this.threshold = FLOOR;
    this.residueHistory = [initialResidue];
  }

  withThreshold(theta: number): this {
    this.threshold = Math.max(theta, FLOOR);
    return this;
  }

  /** Descent rate Delta over the stall window — non-negative by definition. */
  descentRate(): number {
    const h = this.residueHistory;
    if (h.length < 2) return 0;
    const window = Math.min(STALL_WINDOW, h.length - 1);
    const oldest = h[h.length - 1 - window]!;
    const newest = h[h.length - 1]!;
    return Math.max(oldest - newest, 0) / window;
  }

  /**
   * Residue priority P(t) — the scheduling signal.
   *
   * A running task with no committed units yet is UNTRIED (not stalled): it has
   * no descent history to measure, so it receives a positive bootstrap priority
   * proportional to how far it is above threshold, guaranteeing it is dispatched
   * at least once. Only a task that has been tried and then goes flat is stalled
   * (Delta = 0 -> P = 0). This is what makes liveness hold from a cold start.
   */
  priority(): number {
    if (this.state !== "running") return 0;
    if (this.residue <= this.threshold) return Infinity; // finish immediately
    const gap = Math.max(this.residue - this.threshold, FLOOR);
    if (this.trajectoryCount === 0) return gap; // untried -> bootstrap dispatch
    const delta = this.descentRate();
    if (delta === 0) return 0; // tried and flat -> stalled, do not feed
    return delta / gap;
  }

  /** Record a new residue reading after one committed unit; returns priority. */
  commitUnit(newResidue: number): number {
    this.trajectoryCount++; // strictly monotone — never decremented
    this.residue = Math.max(newResidue, FLOOR);
    this.residueHistory.push(this.residue);

    if (this.residue <= this.threshold) {
      this.state = "sufficient";
    } else if (this.descentRate() === 0 && this.residueHistory.length > STALL_WINDOW) {
      const flat = this.residueHistory
        .slice(-STALL_WINDOW)
        .every((r) => Math.abs(r - this.residue) < 1e-9);
      if (flat) this.state = "stalled";
    }
    return this.priority();
  }

  /** Expected termination unit count: n_term(K,d) = 1 + ceil(log_{d+1}(K/d)). */
  terminationBound(): number {
    const k = this.complexity;
    const d = this.opTypes;
    if (d <= 0 || k <= 0) return 1;
    const inner = Math.log(k / d) / Math.log(d + 1);
    return 1 + Math.ceil(inner);
  }

  isBehindCurve(): boolean {
    return this.trajectoryCount > this.terminationBound() && this.state === "running";
  }
}

/** One dispatched unit's result. */
export interface TickResult {
  readonly taskId: string;
  readonly trajectoryCount: number;
  readonly residueBefore: number;
  readonly residueAfter: number;
  readonly priorityAfter: number;
  readonly state: TaskState;
}

/** The residue-driven scheduler. */
export class Scheduler {
  readonly tasks = new Map<string, Task>();

  addTask(task: Task): void {
    this.tasks.set(task.id, task);
  }

  /**
   * One scheduler pass over `budget` dispatches. Repeatedly dispatches the
   * highest-priority running task; a stalled-only frontier (all P=0) declines.
   * `dispatch(task)` performs the work unit and returns the new residue.
   */
  tick(budget: number, dispatch: (task: Task) => number): TickResult[] {
    const results: TickResult[] = [];
    let remaining = budget;
    while (remaining > 0) {
      // pick the highest-priority running task
      let top: Task | undefined;
      let topP = -Infinity;
      for (const t of this.tasks.values()) {
        if (t.state !== "running") continue;
        const p = t.priority();
        if (p > topP) {
          topP = p;
          top = t;
        }
      }
      if (top === undefined) break; // no runnable tasks
      if (topP === 0) break; // all runnable tasks stalled -> decline

      const residueBefore = top.residue;
      const newResidue = dispatch(top);
      const priorityAfter = top.commitUnit(newResidue);
      results.push({
        taskId: top.id,
        trajectoryCount: top.trajectoryCount,
        residueBefore,
        residueAfter: top.residue,
        priorityAfter,
        state: top.state,
      });
      remaining--;
    }
    return results;
  }

  runningCount(): number {
    let n = 0;
    for (const t of this.tasks.values()) if (t.state === "running") n++;
    return n;
  }

  stalledTasks(): Task[] {
    return [...this.tasks.values()].filter((t) => t.state === "stalled");
  }
}
